### PR TITLE
IGNITE-15486 JUnit configuration extension implemented

### PR DESCRIPTION
**Pull request by ibessonov**
https://issues.apache.org/jira/browse/IGNITE-15486

### CODE REVIEW

File: modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtensionTest.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+import java.util.concurrent.ExecutionException;
+import org.apache.ignite.internal.configuration.sample.DiscoveryConfiguration;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+/**
+ * Test basic scenarios of {@link ConfigurationExtension}
```
Comment (tkalkirill) [inline modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtensionTest.java:28]:
```suggestion
 * Test basic scenarios of {@link ConfigurationExtension}.
```

File: modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+            InjectConfiguration.class,
+            field -> supportType(field.getType()),
+            HierarchyTraversalMode.TOP_DOWN
+        );
+
+        return fields;
+    }
+
+    /**
+     * Checks that instance of the gicen class can be injected by the extension.
```
Comment (tkalkirill) [inline modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java:185]:
```suggestion
     * Checks that instance of the given class can be injected by the extension.
```

File: modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+     * Instantiates configuration instance for injection.
+     *
+     * @param type Type of the field or parameter. Class name must end with {@code Configuration}.
+     * @param annotation Annotation present on the field or parameter.
+     * @param cgen Runtime code generator associated with the extension instance.
+     * @return Mock configuration instance.
+     * @throws ClassNotFoundException If corresponding configuration schema class is not found.
+     * @see #supportType(Class)
+     */
+    private static Object cfgValue(Class<?> type, InjectConfiguration annotation, ConfigurationAsmGenerator cgen)
```
Comment (tkalkirill) [inline modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java:117]:
```suggestion
    private static Object cfgValue(
    Class<?> type, 
    InjectConfiguration annotation, 
    ConfigurationAsmGenerator cgen
    )
```

**Review by tkalkirill — APPROVED** [review]

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/DynamicConfigurationChanger.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+package org.apache.ignite.internal.configuration;
+
+import java.util.concurrent.CompletableFuture;
+import org.apache.ignite.configuration.RootKey;
+import org.apache.ignite.internal.configuration.tree.ConfigurationSource;
+import org.apache.ignite.internal.configuration.tree.InnerNode;
+
+/**
+ * Interface to provide configuration access to {@link DynamicConfiguration}, {@link NamedListConfiguration} and
```
Comment (sashapolo) [inline modules/configuration/src/main/java/org/apache/ignite/internal/configuration/DynamicConfigurationChanger.java:26]: `provide configuration access` - what does this sentence mean?

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/DynamicConfigurationChanger.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    /**
+     * Changes the configuration.
+     *
+     * @param source Configuration source to create patch from.
+     * @return Future that is completed on change completion.
+     */
+    CompletableFuture<Void> change(ConfigurationSource source);
+
+    /**
+     * Get root node by root key. Subject to revisiting.
```
Comment (sashapolo) [inline modules/configuration/src/main/java/org/apache/ignite/internal/configuration/DynamicConfigurationChanger.java:39]: Subject to revisiting?

File: modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java (hunk 3 above)
Comment (ibessonov) [reply modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java]: Formatted manually

File: modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+        for (Field field : getMatchingFields(testInstance.getClass())) {
+            field.setAccessible(true);
+
+            field.set(testInstance, cfgValue(field.getType(), field.getAnnotation(InjectConfiguration.class), cgen));
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public void afterEach(ExtensionContext context) throws Exception {
+        context.getStore(ExtensionContext.Namespace.GLOBAL).remove(CGEN_KEY);
```
Comment (sashapolo) [inline modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java:77]: I'm not sure this is the correct namespace to use, as it is shared by all extensions

File: modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+            && supportType(parameterContext.getParameter().getType());
+    }
+
+    /** {@inheritDoc} */
+    @Override public Object resolveParameter(
+        ParameterContext parameterContext, ExtensionContext extensionContext
+    ) throws ParameterResolutionException {
+        Parameter parameter = parameterContext.getParameter();
+
+        var cgen = (ConfigurationAsmGenerator)extensionContext.getStore(ExtensionContext.Namespace.GLOBAL).get(CGEN_KEY);
```
Comment (sashapolo) [inline modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java:94]: there's a `get(Object, Class)` method that you can use here

File: modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+        catch (ClassNotFoundException classNotFoundException) {
+            throw new ParameterResolutionException(
+                "Cannot find a configuration schema class that matches " + parameter.getType().getCanonicalName(),
+                classNotFoundException
+            );
+        }
+    }
+
+    /**
+     * Instantiates configuration instance for injection.
```
Comment (sashapolo) [inline modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java:108]:
```suggestion
     * Instantiates a configuration instance for injection.
```

File: modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+        });
+    }
+
+    /**
+     * Looks for the annotated field inside the given test class.
+     *
+     * @return Annotated fields.
+     */
+    private static List<Field> getMatchingFields(Class<?> testClass) {
+        List<Field> fields = AnnotationSupport.findAnnotatedFields(
```
Comment (sashapolo) [inline modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java:176]:
```suggestion
        return AnnotationSupport.findAnnotatedFields(
```

File: modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java (hunk 7 above)
Comment (sashapolo) [reply modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java:94]: also, this line is too long

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/DynamicConfigurationChanger.java (hunk 5 above)
Comment (ibessonov) [reply modules/configuration/src/main/java/org/apache/ignite/internal/configuration/DynamicConfigurationChanger.java:39]: Removed that line

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/DynamicConfigurationChanger.java (hunk 4 above)
Comment (ibessonov) [reply modules/configuration/src/main/java/org/apache/ignite/internal/configuration/DynamicConfigurationChanger.java:26]: Added a few words

File: modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java (hunk 6 above)
Comment (ibessonov) [reply modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java:77]: done

File: modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java (hunk 7 above)
Comment (ibessonov) [reply modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java:94]: done

File: modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java (hunk 8 above)
Comment (ibessonov) [reply modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java:108]: done

File: modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java (hunk 9 above)
Comment (ibessonov) [reply modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java:176]: sure

**Review by sashapolo — APPROVED** [review]

**Pull request merged**